controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,478 @@
+	if err := r.list(storageClusterPeerList, client.InNamespace(clientMappingConfig.Namespace)); err != nil {
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+	if len(storageClusterPeerList.Items) != 1 {
+		return ctrl.Result{}, fmt.Errorf("expected 1 StorageClusterPeer but got %d", len(storageClusterPeerList.Items))
+	}
+
+	storageClusterPeer := &storageClusterPeerList.Items[0]
+
+	if storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		r.log.Info(
+			"waiting for StorageClusterPeer to be in Peered state",
+			"StorageClusterPeer",
+			storageClusterPeer.Name,
+		)
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	ocsClient, err := providerClient.NewProviderClient(r.ctx, storageClusterPeer.Spec.ApiEndpoint, util.OcsClientTimeout)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRbdMirror(clientMappingConfig, disableMirroring); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		disableMirroring,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		disableMirroring,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if disableMirroring {
+		if controllerutil.RemoveFinalizer(clientMappingConfig, mirroringFinalizer) {
+			r.log.Info("removing finalizer from ConfigMap.")
+			if err := r.update(clientMappingConfig); err != nil {
+				r.log.Info("Failed to remove finalizer from ConfigMap")
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from ConfigMap: %v", err)
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRbdMirror(clientMappingConfig *corev1.ConfigMap, disableMirroring bool) error {
+	rbdMirror := &rookCephv1.CephRBDMirror{}
+	rbdMirror.Name = util.CephRBDMirrorName
+	rbdMirror.Namespace = clientMappingConfig.Namespace
+
+	storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+	if err := r.list(
+		storageConsumers,
+		client.MatchingFields{util.AnnotationIndexName: util.RequestMaintenanceModeAnnotation},
+	); err != nil {
+		return fmt.Errorf("failed to list StorageConsumer(s): %v", err)
+	}
+
+	disableRBDMirror := disableMirroring || len(storageConsumers.Items) >= 1
+
+	if !disableRBDMirror {
+		_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, rbdMirror, func() error {
+			if err := r.own(clientMappingConfig, rbdMirror); err != nil {
+				return err
+			}
+			rbdMirror.Spec.Count = 1
+			return nil
+		})
+		if err != nil {
+			r.log.Error(err, "Failed to create/update the CephRBDMirror", "CephRBDMirror", rbdMirror)
+			return err
+		}
+	} else {
+		if err := r.delete(rbdMirror); err != nil {
+			return fmt.Errorf("failed to delete CephRBDMirror: %v", err)
+		}
+	}
```

Comment:
1. Prefer the positive case for the negative case, it removes negations from the main code path
2. use a variable with a proper name to convey the intent (better than using comments) 

```
	maintenanceModeRequested := len(storageConsumers.Items) >= 1
	if mirroringEnabled || !maintenanceModeRequested {
		// Create or update	
	} else 
		// Delete 
	}
```